src/eigen/lobpcg_hyper_impl.jl

```diff
@@ -254,9 +254,9 @@ end
         BX = B*X
         B_ortho!(X, BX)
     end
-    
+
     n_matvec = M   # Count number of matrix-vector products
-    AX = A*X
+    AX = Matrix(A*X)
```

Comment:
Nope, if I remove it I get the following error in the file `test/compute_jacobian_eigval.jl` when calling lobpcg

```
MethodError: no method matching getindex(::LinearMaps.CompositeMap{Float64, Tuple{LinearMaps.WrappedMap{Float64, Matrix{Float64}}, LinearMaps.FunctionMap{Float64, var"#apply_jacobian#30"{var"#unpack#24"{DataType, PlaneWaveBasis{Float64}}, var"#pack#23"{DataType, PlaneWaveBasis{Float64}}, Vector{Matrix{ComplexF64}}, NamedTuple{(:ham, :basis, :energies, :converged, :ρ, :eigenvalues, :occupation, :εF, :n_iter, :n_ep_extra, :ψ, :diagonalization, :stage), Tuple{Hamiltonian, PlaneWaveBasis{Float64}, Energies{Float64}, Bool, Array{Float64, 4}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Int64, Int64, Vector{Matrix{ComplexF64}}, NamedTuple{(:λ, :X, :residual_norms, :iterations, :converged, :n_matvec), Tuple{Vector{Vector{Float64}}, Vector{Matrix{ComplexF64}}, Vector{Vector{Float64}}, Vector{Int64}, Bool, Int64}}, Symbol}}, PlaneWaveBasis{Float64}}, Nothing}}}, ::Colon, ::Int64)
  Stacktrace:
    [1] maybeview(::LinearMaps.CompositeMap{Float64, Tuple{LinearMaps.WrappedMap{Float64, Matrix{Float64}}, LinearMaps.FunctionMap{Float64, var"#apply_jacobian#30"{var"#unpack#24"{DataType, PlaneWaveBasis{Float64}}, var"#pack#23"{DataType, PlaneWaveBasis{Float64}}, Vector{Matrix{ComplexF64}}, NamedTuple{(:ham, :basis, :energies, :converged, :ρ, :eigenvalues, :occupation, :εF, :n_iter, :n_ep_extra, :ψ, :diagonalization, :stage), Tuple{Hamiltonian, PlaneWaveBasis{Float64}, Energies{Float64}, Bool, Array{Float64, 4}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Int64, Int64, Vector{Matrix{ComplexF64}}, NamedTuple{(:λ, :X, :residual_norms, :iterations, :converged, :n_matvec), Tuple{Vector{Vector{Float64}}, Vector{Matrix{ComplexF64}}, Vector{Vector{Float64}}, Vector{Int64}, Bool, Int64}}, Symbol}}, PlaneWaveBasis{Float64}}, Nothing}}}, ::Function, ::Int64)
      @ Base ./views.jl:132
    [2] (::DFTK.var"#853#855")(n::Int64)
      @ DFTK ./none:0
    [3] iterate
      @ ./generator.jl:47 [inlined]
    [4] collect(itr::Base.Generator{UnitRange{Int64}, DFTK.var"#853#855"})
      @ Base ./array.jl:678
    [5] macro expansion
      @ ~/Documents/Nextcloud/CERMICS/codes/DFTK/DFTK.jl/src/eigen/lobpcg_hyper_impl.jl:278 [inlined]
    [6] (::DFTK.var"#852#854"{Int64, Float64, Bool, LinearMaps.FunctionMap{Float64, var"#apply_jacobian#30"{var"#unpack#24"{DataType, PlaneWaveBasis{Float64}}, var"#pack#23"{DataType, PlaneWaveBasis{Float64}}, Vector{Matrix{ComplexF64}}, NamedTuple{(:ham, :basis, :energies, :converged, :ρ, :eigenvalues, :occupation, :εF, :n_iter, :n_ep_extra, :ψ, :diagonalization, :stage), Tuple{Hamiltonian, PlaneWaveBasis{Float64}, Energies{Float64}, Bool, Array{Float64, 4}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Int64, Int64, Vector{Matrix{ComplexF64}}, NamedTuple{(:λ, :X, :residual_norms, :iterations, :converged, :n_matvec), Tuple{Vector{Vector{Float64}}, Vector{Matrix{ComplexF64}}, Vector{Vector{Float64}}, Vector{Int64}, Bool, Int64}}, Symbol}}, PlaneWaveBasis{Float64}}, Nothing}, UniformScaling{Bool}, FunctionPreconditioner, Float64, Int64})()
      @ DFTK ~/.julia/packages/TimerOutputs/ZmKD7/src/TimerOutput.jl:237
    [7] timeit(f::DFTK.var"#852#854"{Int64, Float64, Bool, LinearMaps.FunctionMap{Float64, var"#apply_jacobian#30"{var"#unpack#24"{DataType, PlaneWaveBasis{Float64}}, var"#pack#23"{DataType, PlaneWaveBasis{Float64}}, Vector{Matrix{ComplexF64}}, NamedTuple{(:ham, :basis, :energies, :converged, :ρ, :eigenvalues, :occupation, :εF, :n_iter, :n_ep_extra, :ψ, :diagonalization, :stage), Tuple{Hamiltonian, PlaneWaveBasis{Float64}, Energies{Float64}, Bool, Array{Float64, 4}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Int64, Int64, Vector{Matrix{ComplexF64}}, NamedTuple{(:λ, :X, :residual_norms, :iterations, :converged, :n_matvec), Tuple{Vector{Vector{Float64}}, Vector{Matrix{ComplexF64}}, Vector{Vector{Float64}}, Vector{Int64}, Bool, Int64}}, Symbol}}, PlaneWaveBasis{Float64}}, Nothing}, UniformScaling{Bool}, FunctionPreconditioner, Float64, Int64}, to::TimerOutputs.TimerOutput, label::String)
      @ TimerOutputs ~/.julia/packages/TimerOutputs/ZmKD7/src/TimerOutput.jl:285
    [8] #LOBPCG#851
      @ ~/.julia/packages/TimerOutputs/ZmKD7/src/TimerOutput.jl:236 [inlined]
    [9] lobpcg_hyper(A::LinearMaps.FunctionMap{Float64, var"#apply_jacobian#30"{var"#unpack#24"{DataType, PlaneWaveBasis{Float64}}, var"#pack#23"{DataType, PlaneWaveBasis{Float64}}, Vector{Matrix{ComplexF64}}, NamedTuple{(:ham, :basis, :energies, :converged, :ρ, :eigenvalues, :occupation, :εF, :n_iter, :n_ep_extra, :ψ, :diagonalization, :stage), Tuple{Hamiltonian, PlaneWaveBasis{Float64}, Energies{Float64}, Bool, Array{Float64, 4}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Int64, Int64, Vector{Matrix{ComplexF64}}, NamedTuple{(:λ, :X, :residual_norms, :iterations, :converged, :n_matvec), Tuple{Vector{Vector{Float64}}, Vector{Matrix{ComplexF64}}, Vector{Vector{Float64}}, Vector{Int64}, Bool, Int64}}, Symbol}}, PlaneWaveBasis{Float64}}, Nothing}, X0::Matrix{Float64}; maxiter::Int64, prec::FunctionPreconditioner, tol::Float64, largest::Bool, n_conv_check::Nothing, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
      @ DFTK ~/Documents/Nextcloud/CERMICS/codes/DFTK/DFTK.jl/src/eigen/diag_lobpcg_hyper.jl:9
   [10] macro expansion
      @ ~/Documents/Nextcloud/CERMICS/codes/DFTK/DFTK.jl/test/compute_jacobian_eigval.jl:61 [inlined]
   [11] macro expansion
      @ /build/julia/src/julia-1.6.1/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1151 [inlined]
   [12] top-level scope
      @ ~/Documents/Nextcloud/CERMICS/codes/DFTK/DFTK.jl/test/compute_jacobian_eigval.jl:12
   [13] include(fname::String)
      @ Base.MainInclude ./client.jl:444
   [14] top-level scope
      @ REPL[9]:1
```

Reply:
We definitely should not change it here because we don't want an extra alloc. There's something wrong with the linear map - possibly it does something fancy for A * X like make a composite map instead of just doing A(x) column by column?

Reply:
yup: https://jutho.github.io/LinearMaps.jl/dev/types/#Base.:*-Tuple{AbstractArray{T,2}%20where%20T,LinearMap}

Reply:
Hm, this is unfortunate. I don't have a great solution here, so we should just define a custom type, a la FunctionPreconditioner.

Reply:
See https://github.com/Jutho/LinearMaps.jl/issues/156

Reply:
Ok, I'll focus on this after finishing the review.

Reply:
So do we follow the solution they suggest ( `convert(AbstractMatrix, A * X)` ) or do we define a custom type ?

Reply:
Let's see where the conversation ends up, but maybe the convert is not so bad. Or just `AX = similar(X); mul!(AX, A, X)` and the same for B above?